### PR TITLE
[ChatUpdate] Added SentChatMessage-History & Chat Tab-Completion

### PR DESCRIPTION
**Pull request by iZePlayzYT**
- Fixed a few bugs and bad code regarding "default commands"
- Added SentChatMessages-History (including navigation)
- Added Tab-Completion for Main-Commands
- Added Tab-Completion for Sub-Commands
- Added Tab-Completion for Playernames (Not working correctly yet. Infos -> Check the comment at line 425 in the class "djui_chat_box.c" (Method: djui_chat_box_input_on_key_down, "case SCANCODE_TAB:")

### CODE REVIEW

**Comment by yoyeet961** [comment]
you should make this pull request on the dev branch of https://github.com/sm64ex-coop-dev/sm64ex-coop

**Comment by yoyeet961** [comment]
tho its a private repo tho so you should ping djoslin0 in the coop server asking to get access to it and your github username

**Comment by yoyeet961** [comment]
also why did you increase the hooked chat command limit to 8192? hooking so much chat commands might even crash the game

**Review by yoyeet961 — COMMENTED** [review]
when i tried using printf, the game always crashed, so consider not using printf.

File: src/pc/lua/smlua_hooks.c [hunk 1]
```diff
@@ -1304,7 +1311,7 @@ struct LuaHookedChatCommand {
     struct Mod* mod;
 };
 
-#define MAX_HOOKED_CHAT_COMMANDS 64
+#define MAX_HOOKED_CHAT_COMMANDS 8192
```
Comment (yoyeet961) [inline src/pc/lua/smlua_hooks.c]: you should change this back to 64, what mod needs 8192 commands

File: src/pc/djui/djui_chat_box.c [hunk 2]
```diff
@@ -9,6 +10,98 @@ struct DjuiChatBox* gDjuiChatBox = NULL;
 bool gDjuiChatBoxFocus = false;
 static bool sDjuiChatBoxClearText = false;
 
+#define MAX_HISTORY_SIZE 1000
```
Comment (yoyeet961) [inline src/pc/djui/djui_chat_box.c]: who will press arrow up 1000 times?

File: src/pc/djui/djui_chat_box.c [hunk 3]
```diff
@@ -9,6 +10,98 @@ struct DjuiChatBox* gDjuiChatBox = NULL;
...
+//}
+
+//void resetTabCompletionAll(void) {
+//    resetTabCompletionCommands();
+//    resetTabCompletionPlayernames();
+//}
+
+void resetTabCompletion(void) {
+    iTabCompletionIndex = -1;
+    strncpy(sTabCompletionOriginalText, "", MAX_MSG_LENGTH - 1);
```
Comment (yoyeet961) [inline src/pc/djui/djui_chat_box.c]: consider using snprintf instead of strncpy

File: src/pc/djui/djui_chat_box.c [hunk 4]
```diff
@@ -62,7 +159,142 @@ static void djui_chat_box_input_escape(struct DjuiInputbox* chatInput) {
...
+                }
+                currentIndex++;
+            }
+        }
+    }
+
+    for (int i = 0; players[i] != NULL; i++) {
+        free(players[i]);
+    }
+    free(players);
```
Comment (yoyeet961) [inline src/pc/djui/djui_chat_box.c]: why are you doing free()? doesnt that deallocate the return value of smlua_get_chat_player_list()?

File: src/pc/djui/djui_chat_box.c [hunk 5]
```diff
@@ -1,4 +1,5 @@
 #include <stdio.h>
+#include <stdlib.h>
 #include <string.h>
 #include "pc/network/network.h"
 #include "pc/lua/smlua_hooks.h"
```
Comment (yoyeet961) [inline src/pc/djui/djui_chat_box.c:5]: you need to include djui_interactable.h or else there wont be any SCANCODE_ stuff

**Review by yoyeet961 — CHANGES_REQUESTED** [review]
there are a lot of stuff in your code that needs to be modified/added

File: src/pc/djui/djui_chat_box.c (hunk 2 above)
Comment (iZePlayzYT) [reply src/pc/djui/djui_chat_box.c]: i decreased it to 256

File: src/pc/djui/djui_chat_box.c (hunk 3 above)
Comment (iZePlayzYT) [reply src/pc/djui/djui_chat_box.c]: done

File: src/pc/djui/djui_chat_box.c (hunk 4 above)
Comment (iZePlayzYT) [reply src/pc/djui/djui_chat_box.c]: no, it does not

File: src/pc/djui/djui_chat_box.c (hunk 5 above)
Comment (iZePlayzYT) [reply src/pc/djui/djui_chat_box.c:5]: never touch a running system or code.

File: src/pc/lua/smlua_hooks.c (hunk 1 above)
Comment (iZePlayzYT) [reply src/pc/lua/smlua_hooks.c]: i decreased it to 512

**Comment by iZePlayzYT** [comment]
some changes done

**Comment by yoyeet961** [comment]
ok but you still didnt include djui_interactables.h

File: src/pc/djui/djui_chat_box.c (hunk 5 above)
Comment (yoyeet961) [reply src/pc/djui/djui_chat_box.c:5]: what
if you don't include it then all the SCANCODE_ stuff will be NULL, and so all of it will pretty much be 0

**Review by iZePlayzYT — COMMENTED** [review]
test123

**Pull request closed without merging**